File: backend/starships.py
"""This program pulls data from the Star Wars API (http://swapi.dev/) 
and write to stdout a JSON object containing a list of starship
names and for each starship, a list of pilot names who have used that 
starship.
"""

import urllib.request
import urllib.parse
import json
from multiprocessing.dummy import Pool


class Starships:
    """Collect data about all available starships and their pilots and 
    output the formatted JSON to stdout.
    """

    def __init__(self):
        self.api_url = 'https://swapi.dev/api/'
        self.pilot_urls = set()
        self.pilot_info = {}
        self.output_data = {'starships': []}

    def _api_request(self, url):
        """Query the URL specified and return the result as a 
        Python dict.
        """
        request = urllib.request.Request(url)
        response = urllib.request.urlopen(request)
        return json.loads(response.read().decode(response.info().get_param(
            'charset') or 'utf-8'))

    def _get_starships(self):
        """Download all starship names and associated pilot URLs from 
        the /starships endpoint. 
        """
        next_url = self.api_url + 'starships/'
        while next_url:
            starships_data = self._api_request(next_url)
            for starship in starships_data['results']:
                self.output_data['starships'].append({'name': starship['name'], 
                    'pilots': starship['pilots']})
                for pilot in starship['pilots']:
                    self.pilot_urls.add(pilot)
            next_url = starships_data['next']   

    def _get_pilot_name(self, pilot_url):
        """Find a single pilot's name from the URL.  
        """
        pilot_data = self._api_request(pilot_url)
        self.pilot_info[pilot_url] = pilot_data['name']

    def _get_all_pilot_names(self):
        """Find all pilot names based on the URLs, using 
        multithreading. 
        """
        pool = Pool()
        results = pool.map(self._get_pilot_name, self.pilot_urls)
        pool.close()
        pool.join()

    def _update_pilot_info(self):
        """Replace each pilot URL in the output data with its 
        associated name. 
        """
        for starship in self.output_data['starships']:
            for pilot_index in range(len(starship['pilots'])):
                pilot_url = starship['pilots'][pilot_index]
                pilot_name = self.pilot_info[pilot_url]
                starship['pilots'][pilot_index] = pilot_name        

    def main(self):
        """Run all steps and print the formatted JSON output to the 
        console.
        """
        self._get_starships()
        self._get_all_pilot_names()
        self._update_pilot_info()
        print(json.dumps(self.output_data, indent=2))


if __name__ == '__main__':
  Starships().main()

File: frontend/star_wars_api/src/starships.js
import React from "react";
import {Error} from "./error";
import "./starships.css";


export class Starships extends React.Component {
  // Pulls a list of all available starships from the Star Wars API,
  // then pulls the pilot name for each pilot URL and loads a dropdown with 
  // each starship name.
  // When a starship is selected, loads a table of all pilot names for that 
  // starship.  
  constructor(props) {
    super(props);
    this.state = {error: null, downloadStarships: false, 
      downloadPilots: false, loadPilotNames: false, starshipName: null, 
      starshipDropdown: null, 
      pilotNameRows: <tr><td>Loading...</td></tr>};
    this.nextUrl = "https://swapi.dev/api/starships/";
    this.starshipData = {};
    this.pilotUrls = new Set();
    this.pilotInfo = {};
    this.selectStarshipName = this.selectStarshipName.bind(this);
  }
  getStarships() { 
    // Download all starship names and associated pilot URLs from the 
    // /starships endpoint.
    if (this.nextUrl != null) {
      fetch(this.nextUrl)
        .then(response => response.json())
        .then(
          (result) => {
            for (const starship of result["results"]) {
              (this.starshipData[starship["name"]] = 
                starship["pilots"]);
              for (const pilotUrl of starship["pilots"]) {
                this.pilotUrls.add(pilotUrl);
              }
            }
            this.nextUrl = result["next"];
            if (this.nextUrl != null) {
              this.setState({
                downloadStarships: true
              });
            }
            else {
              this.setState({
                downloadStarships: false,
                downloadPilots: true
              });
            }
          },
          (error) => {
            this.setState({
              error: error.message,
              downloadStarships: false
            });
          }
        )
    }
  }
  getPilotNames() {
    // Find the pilot name associated with each pilot URL.
    const allPilotUrls = this.pilotUrls.values();
    for (const pilotUrl of allPilotUrls) {
      fetch(pilotUrl)
        .then(response => response.json())
        .then(
          (result) => {
            this.pilotInfo[pilotUrl] = result["name"];
            this.pilotUrls.delete(pilotUrl);
            if (this.pilotUrls.size === 0) {
              this.setState({
                downloadPilots: false
              });
            }
          },
          (error) => {
            this.setState({
              error: error.message,
              downloadPilots: false
            });
            this.pilotInfo[pilotUrl] = "Unknown";
            this.pilotUrls.delete(pilotUrl);
          }
      )
    }
  }
  updatePilotInfo() {
    // Replace each pilot URL in the starship data with its associated name.
    for (const pilotEntries of Object.values(this.starshipData)) {
      for (let i = 0; i < pilotEntries.length; i++) {
        const pilotUrl = pilotEntries[i];
        const pilotName = this.pilotInfo[pilotUrl];
        pilotEntries[i] = pilotName;
      }
    }
  }
  loadStarshipDropdown() {
    // Load all the starship names into the form dropdown.
    let starshipDropdown = [];
    for (const starshipName of Object.keys(this.starshipData)) {
      starshipDropdown.push(
        <option value={starshipName} key={starshipName}>{starshipName}
        </option>
      );
    }
    this.setState({
      starshipDropdown: starshipDropdown,
      pilotNameRows: null
    });
  }
  selectStarshipName(event) {
    // Handle changes made in the Starship Name select dropdown.
    const target = event.target;
    const value = target.value;
    this.setState({
      starshipName: value,
      loadPilotNames: true
    });
  }
  loadPilotNames() {
    // Display all pilot names associated with the selected starship name 
    // in a table. 
    let pilotNameRows = [];
    if (this.starshipData[this.state.starshipName].length === 0) {
      pilotNameRows.push(
        <tr key="1"><td>No pilots found!</td></tr>
      );
    }
    for (const pilotName of this.starshipData[this.state.starshipName]) {
      pilotNameRows.push(
        <tr key={pilotName}><td>{pilotName}</td></tr>
      );
    }
    this.setState({
      pilotNameRows: pilotNameRows,
      loadPilotNames: false
    });
  }
  componentDidMount() {
    this.getStarships();
  }
  componentDidUpdate() {
    if (this.state.downloadStarships) {
      this.getStarships();
    }
    if (this.state.downloadPilots) {
      this.getPilotNames();
    }
    if (this.pilotUrls.size === 0 && this.state.starshipDropdown === null) {
      this.updatePilotInfo();
      this.loadStarshipDropdown();
    }
    if (this.state.loadPilotNames) {
      this.loadPilotNames();
    }
  }
  render() {
    return (
      <React.Fragment>
      <p className="title">Star Wars Starship Pilots</p>
      <form className="starship-select">
      <p><label htmlFor="starshipName">Starship:&nbsp;
      <select name="starshipName" onChange={this.selectStarshipName}>
        <option value=""></option>
        {this.state.starshipDropdown}
      </select>
      </label></p>
      </form>
      <table className="pilot-table">
        <tbody>
        {this.state.pilotNameRows}
        </tbody>
      </table>
      {this.state.error && <Error message={this.state.error} />}
      </React.Fragment>
    );
  }
}